pkg/wal/processor/transformer/wal_transformer_validator.go

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package transformer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v5/pgtype"
+	pglib "github.com/xataio/pgstream/internal/postgres"
+	"github.com/xataio/pgstream/pkg/transformers"
+	"golang.org/x/exp/slices"
+)
+
+func GetTransformerValidationFunc(ctx context.Context, pgurl, snapshoturl string) func(*Transformer) error {
+	// if pgurl or snapshot url is provided, use postgres validator
+	if pgurl == "" {
+		if snapshoturl == "" {
+			return nil
+		}
+		pgurl = snapshoturl
+	}
+	return pgTransformerValidator(ctx, pgurl)
+}
+
+func pgTransformerValidator(ctx context.Context, pgurl string) func(t *Transformer) error {
+	return func(t *Transformer) error {
+		conn, err := pglib.NewConn(ctx, pgurl)
+		if err != nil {
+			return fmt.Errorf("creating postgres connection pool: %w", err)
+		}
+		defer conn.Close(context.Background())
+
+		transformerMap := t.GetTransformerMap()
+		for schemaTable, columnTransformers := range transformerMap {
+			query := fmt.Sprintf("SELECT * FROM %s LIMIT 0", schemaTable)
+			rows, err := conn.Query(ctx, query)
+			if err != nil {
+				return fmt.Errorf("querying table rows: %w", err)
+			}
+			defer rows.Close()
+			fieldDescriptions := rows.FieldDescriptions()
+
+			// TODO: maybe error out if len(fieldDescriptions) != len(columnTransformers)
+			// if we start requiring a transformer for every column (noop transformers)
+
+			// map column names to column pg type OIDs, skip columns that don't have a transformer
+			mappedColumns := make(map[string]uint32, len(fieldDescriptions))
```

Comment:
Do we need 2 loops here, or can it be done in one? 

```
        // check that all column transformers are compatible with corresponding column types
	for _, desc := range fieldDescriptions {
	        tr, found := columnTransformers[desc.Name]
		if !found {
			continue
		}
		if !pgTypeCompatibleWithTransformerType(tr.CompatibleTypes(),desc.DataTypeOID) {
			return fmt.Errorf("transformer specified for column %s does not support pg data type", colName)
		}
	}
```

Reply:
it could be done in a single loop but I also wanted to validate the rules do not have any non-existing column.

Reply:
Maybe add a comment to be explicit about what this is validating against.